docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,249 @@
+
+The functionality we want to expose via this api is:
+
+- `CollectBundle(spec{}) (bundlePath string, error)`
+  - collect a support bundle from a spec.
+  - takes a parsed spec struct as an parameter.
+  - returns a path to the bundle directory and errors.
+  - to minimise IO and increase collection speed. redactors should be run inline, redacting data in memory before it's saved to a bundle.
+- `RedactBundle(bundlePath string) error`
+  - redact an already collected bundle, takes a path to the bundle as an parameter.
+  - returns any errors
+- `AnalyzeBundle(spec{}, bundlePath string) (results{},error)`
+  - run analysers from the spec and return the analysis results struct
+  - returns analysis struct and errors
+- `ArchiveBundle(bundlePath string, compressionMethod string, destination string) error`
+  - generates a tar archive of the bundle directory at the specified path with optional compression
+  - takes bundle path, compression method and destination as parameters.
+  - returns errors
+- `ParseSpecFile(specFile string) (spec{},error)`
+  - parses a spec file and returns an unmarshalled troubleshoot spec struct.
+  - could be used for linting
+- `ServeBundle(bundlePath string, kubeconfig string, port int)`
+  - starts a sbctl server using the specified bundle and port, outputting a kubeconfig at a specified location.
+
+### Usage patterns
+
+- generate a support bundle
+
+  `troubleshoot collect supportbundle.yaml`
+
+  `troubleshoot collect supportbundle.yaml secrets/default/kotsadm-appslug-supportbundle`
+
+  `troubleshoot collect https://kots.io`
+
+  in the troubleshoot cli code this could look like:
+
+```
+spec,_ := troubleshoot.ParseSpecFile("spec.yaml")
+bundleFile,_ := troubleshoot.CollectBundle(spec)
+analysisResults,_ := troubleshoot.AnalyzeBundle(spec,bundleFile)
+```
+
+- use a spec to return a go/no-go preflight outcome
+
+  `troubleshoot preflight spec.yaml`
+
+  This should not only clearly state any reasons for failing, but also use standardised exit codes that can be used by automation tools.
+
+  We should also continue to accept standard in, accepting any preflights from a yaml multidoc.
+
+  `helm template | troubleshoot preflight -`
+
+- Interact with an existing support bundle with kubectl.
+
+  `troubleshoot inspect support-bundle-12-12-2001.tar.gz`
+
+  This should set the `KUBECONFIG` environment variable, as well as a `TROUBLESHOOT_SHELL` environment variable before spawning a subshell.
+  This is to allow prompts and shell environments to be able to detect that they're running in a troubleshoot spawned shell, much like tmux et-al.
+
+  `troubleshoot inspect --non-interactive support-bundle-12-12-2001.tar.gz`
+
+  this should behave much like the `sbctl serve` command does today, printing the path to it's kubeconfig location to stdout or optionally take a flag to specify the kubeconfig location.
```

Comment:
I don't follow what you mean by the flag to specify kubeconfig location - doesn't inspect make a temporary kubeconfig, and we'd not want to overwrite `.kube/config`?

does `--non-interactive` mean "no shell please"?  Maybe want to skip that from the first example above.

Reply:
> does --non-interactive mean "no shell please"?

that's how I interpret it 👍

Reply:
> this should behave much like the `sbctl serve` command does today, printing the path to it's kubeconfig location to stdout or optionally take a flag to specify the kubeconfig location.

the concept here makes sense to me 👍 either make a temp kubeconfig and output an `export KUBECONFIG=...` var, or allow a CLI arg to say where to write it out, and then output the `export KUBECONFIG=specified-location` instead

Reply:
yeh so my thoughts here are:
- if you're a human running `kubectl inspect` it's very likely that you're going to want to run kubectl to interact with the bundle. thus defaulting to a shell provisioned with the suitable env-vars removes complexity for people just wanting something to work. 
- if you're writing automation around `kubectl inspect` you might not want it to spawn a nested shell, and extracting the kubeconfig from stdout generates extra handling for scripts. being able to tell troubleshoot where you want to store the kubeconfig allows us to guarantee a location for the kubeconfig that can be used for other tooling in a script / automation pipeline. (i.e a github action)

Reply:
yep, both use cases are valid 👍